File: app/src/main/java/com/example/konye/lingo/api/ApiService.java
package com.example.konye.lingo.api;

import com.example.konye.lingo.models.ChangePassword;
import com.example.konye.lingo.models.CreateUser;
import com.example.konye.lingo.models.AuthResponse;
import com.example.konye.lingo.models.LoginUser;
import com.example.konye.lingo.models.PasswordRequest;
import com.example.konye.lingo.models.PasswordResponse;
import com.example.konye.lingo.models.UpdateUser;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface ApiService {
    String BASEURL = "http://hamdanae.com/api/v1/";

    //Auth Endpoints
    @POST("create")
    Observable<Response<AuthResponse>> registerUser(@Body CreateUser user);

    @POST("update")
    Observable<Response<AuthResponse>> updateUser(@Header("MAHADUM-TOKEN") String token,
                                                  @Body UpdateUser user);

    @POST("login")
    Observable<Response<AuthResponse>> login(@Body LoginUser login);

    @POST("validatepw")
    Observable<Response<PasswordResponse>> validatePassword(@Header("MAHADUM-TOKEN") String token,
                                                            @Body PasswordRequest password);

    @POST("changepw")
    Observable<Response<PasswordResponse>> changePassword(@Header("MAHADUM-TOKEN") String token,
                                                          @Body ChangePassword change);

    //Parent Endpoints
    @POST("parent/billing")
    Call<ResponseBody> addPayment(@Header("MAHADUM-TOKEN") String token);

    @GET("parent/kids")
    Call<ResponseBody> getKidsDetails(@Header("MAHADUM-TOKEN") String token);

    @POST("parent/kids/add")
    Call<ResponseBody> addKid(@Header("MAHADUM-TOKEN") String token);

    //Teacher Endpoints
    @POST("teacher/school")
    Call<ResponseBody> teacherToSchool(@Header("MAHADUM-TOKEN") String token);

    @GET("teacher/kids")
    Call<ResponseBody> kidsInSchoolAndCourse(@Header("MAHADUM-TOKEN") String token);

    //Course Endpoints
    @GET("courses/all")
    Call<ResponseBody> getAllCourses(@Header("MAHADUM-TOKEN") String token);

    @GET("courses/{course-id}/summary")
    Call<ResponseBody> getCourseSummary(@Header("MAHADUM-TOKEN") String token,
                                        @Path("course-id") int id);

    @GET("courses/{course-id}/details")
    Call<ResponseBody> getCourseDetails(@Header("MAHADUM-TOKEN") String token,
                                        @Path("course-id") int id);

    //Lesson Endpoints
    @GET("lessons/all")
    Call<ResponseBody> getAllLessons(@Header("MAHADUM-TOKEN") String token);

    @GET("lessons/{lesson-id}")
    Call<ResponseBody> getLessonById(@Header("MAHADUM-TOKEN") String token,
                                     @Path("lesson-id") int id);

    //Kids Endpoints
    @POST("kids/register/{course-id}")
    Call<ResponseBody> registerKidToCourse(@Header("MAHADUM-TOKEN") String token,
                                           @Path("course-id") int id);

    @POST("kids/start/{course-id}")
    Call<ResponseBody> startCourse(@Header("MAHADUM-TOKEN") String token,
                                   @Path("course-id") int id);

    @GET("kids/status/{course-id}")
    Call<ResponseBody> courseProgress(@Header("MAHADUM-TOKEN") String token,
                                      @Path("course-id") int id);
}

File: app/src/main/java/com/mahadum360/mahadum/data/ApiService.kt
package com.mahadum360.mahadum.data


import com.mahadum360.mahadum.models.*

import io.reactivex.Observable
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST
import retrofit2.http.Path

interface ApiService {

    //Auth Endpoints
    @POST("create")
    fun registerUser(@Body user: CreateUser): Observable<Response<AuthResponse>>

    @POST("update")
    fun updateUser(@Header("MAHADUM-TOKEN") token: String,
                   @Body user: UpdateUser): Observable<Response<AuthResponse>>

    @POST("login")
    fun login(@Body login: LoginUser): Observable<Response<AuthResponse>>

    @POST("validatepw")
    fun validatePassword(@Header("MAHADUM-TOKEN") token: String,
                         @Body password: PasswordRequest): Observable<Response<PasswordResponse>>

    @POST("changepw")
    fun changePassword(@Header("MAHADUM-TOKEN") token: String,
                       @Body change: ChangePassword): Observable<Response<PasswordResponse>>

    //Parent Endpoints
    @POST("parent/billing")
    fun addPayment(@Header("MAHADUM-TOKEN") token: String): Call<ResponseBody>

    @GET("parent/kids")
    fun getKidsDetails(@Header("MAHADUM-TOKEN") token: String): Call<ResponseBody>

    @POST("parent/addkid")
    fun addKid(@Header("MAHADUM-TOKEN") token: String,
               @Body addKid: AddKid): Observable<Response<StatusResponse>>

    //Teacher Endpoints
    @POST("teacher/school")
    fun teacherToSchool(@Header("MAHADUM-TOKEN") token: String): Call<ResponseBody>

    @GET("teacher/kids")
    fun kidsInSchoolAndCourse(@Header("MAHADUM-TOKEN") token: String): Call<ResponseBody>

    //Course Endpoints
    @GET("courses/all")
    fun getAllCourses(@Header("MAHADUM-TOKEN") token: String): Call<ResponseBody>

    @GET("courses/{course-id}/summary")
    fun getCourseSummary(@Header("MAHADUM-TOKEN") token: String,
                         @Path("course-id") id: Int): Call<ResponseBody>

    @GET("courses/{course-id}/details")
    fun getCourseDetails(@Header("MAHADUM-TOKEN") token: String,
                         @Path("course-id") id: Int): Call<ResponseBody>

    //Lesson Endpoints
    @GET("lessons/all")
    fun getAllLessons(@Header("MAHADUM-TOKEN") token: String): Call<ResponseBody>

    @GET("lessons/{lesson-id}")
    fun getLessonById(@Header("MAHADUM-TOKEN") token: String,
                      @Path("lesson-id") id: Int): Call<ResponseBody>

    //Kids Endpoints
    @POST("kids/register/{course-id}")
    fun registerKidToCourse(@Header("MAHADUM-TOKEN") token: String,
                            @Path("course-id") id: Int): Call<ResponseBody>

    @POST("kids/start/{course-id}")
    fun startCourse(@Header("MAHADUM-TOKEN") token: String,
                    @Path("course-id") id: Int): Call<ResponseBody>

    @GET("kids/status/{course-id}")
    fun courseProgress(@Header("MAHADUM-TOKEN") token: String,
                       @Path("course-id") id: Int): Call<ResponseBody>

    companion object {
        val BASEURL = "http://hamdanae.com/api/v1/"
    }
}
